File: project/module-compat-bukkit/src/main/java/cn/fd/ratziel/compat/inject/IntrusiveClassLoader.java
package cn.fd.ratziel.compat.inject;

import taboolib.common.classloader.IsolatedClassLoader;

/**
 * IntrusiveClassLoader
 *
 * @author TheFloodDragon
 * @since 2024/3/23 22:10
 */
public class IntrusiveClassLoader extends ClassLoader {

    IntrusiveClassLoader(ClassLoader parent) {
        super(parent);
    }

    @Override
    public Class<?> loadClass(String name) throws ClassNotFoundException {
        return this.loadClass(name, false);
    }

    @Override
    public Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
        synchronized (getClassLoadingLock(name)) {
            // 优先父级加载
            Class<?> find = loadClassOrNull(getParent(), name);
            // 隔离类加载器加载 (不检查其父级)
            if (find == null) try {
                if (name.startsWith("cn.fd.ratziel")) {
                    find = IsolatedClassLoader.INSTANCE.loadClass(name, resolve, false);
                }
            } catch (ClassNotFoundException ignored) {
            }
            // 检查结果
            if (find == null) throw new ClassNotFoundException();
            // 返回值
            return find;
        }
    }

    public static Class<?> loadClassOrNull(ClassLoader loader, String name) {
        try {
            return loader.loadClass(name);
        } catch (ClassNotFoundException ignored) {
            return null;
        }
    }

}

File: project/module-compat-bukkit/src/main/kotlin/cn/fd/ratziel/compat/inject/IntrusiveClassLoader.kt
package cn.fd.ratziel.compat.inject

import cn.fd.ratziel.core.function.loadClassOrNull
import taboolib.common.classloader.IsolatedClassLoader

/**
 * IntrusiveClassLoader
 *
 * @author TheFloodDragon
 * @since 2024/3/22 21:27
 */
class IntrusiveClassLoader(parent: ClassLoader) : ClassLoader(parent) {

    override fun loadClass(name: String?) = loadClass(name, false)

    override fun loadClass(name: String?, resolve: Boolean): Class<*> {
        // 优先父级加载
        var find = parent.loadClassOrNull(name)
        // 隔离类加载器加载 (不检查其父级)
        if (find == null) find = try {
            if (name?.startsWith("cn.fd.ratziel") == true)
                IsolatedClassLoader.INSTANCE.loadClass(name, resolve, false)
            else null
        } catch (_: ClassNotFoundException) {
            null
        }
        // 返回值
        return find ?: throw ClassNotFoundException()
    }

}